server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,416 @@
+            assert maxNumberOfRetries >= 0;
+            this.maxNumberOfRetries = maxNumberOfRetries;
+            return this;
+        }
+
+        /**
+         * Builds a new bulk processor.
+         */
+        public BulkProcessor2 build() {
+            return new BulkProcessor2(
+                consumer,
+                maxNumberOfRetries,
+                listener,
+                concurrentRequests,
+                maxRequestsInBulk,
+                maxBulkSizeInBytes,
+                maxBulkRequestQueueSize,
+                queuePollingInterval,
+                flushInterval,
+                flushScheduler,
+                retryScheduler,
+                onClose
+            );
+        }
+
+    }
+
+    /**
+     * @param consumer The consumer that is called to fulfil bulk operations
+     * @param listener The BulkProcessor2 listener that gets called on bulk events
+     * @param name     The name of this processor, e.g. to identify the scheduler threads
+     * @return the builder for BulkProcessor2
+     */
+    public static Builder builder(BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer, Listener listener, String name) {
+        Objects.requireNonNull(consumer, "consumer");
+        Objects.requireNonNull(listener, "listener");
+        final ScheduledThreadPoolExecutor flushScheduler = Scheduler.initScheduler(Settings.EMPTY, name + FLUSH_SCHEDULER_NAME_SUFFIX);
+        final ScheduledThreadPoolExecutor retryScheduler = Scheduler.initScheduler(Settings.EMPTY, name + RETRY_SCHEDULER_NAME_SUFFIX);
+        return new Builder(consumer, listener, buildScheduler(flushScheduler), buildScheduler(retryScheduler), () -> {
+            Scheduler.terminate(flushScheduler, 10, TimeUnit.SECONDS);
+            Scheduler.terminate(retryScheduler, 10, TimeUnit.SECONDS);
+        });
+    }
+
+    private static Scheduler buildScheduler(ScheduledThreadPoolExecutor scheduledThreadPoolExecutor) {
+        return (command, delay, executor) -> Scheduler.wrapAsScheduledCancellable(
+            scheduledThreadPoolExecutor.schedule(command, delay.millis(), TimeUnit.MILLISECONDS)
+        );
+    }
+
+    private final int bulkActions;
+    private final long bulkSize;
+
+    private final Scheduler.Cancellable cancellableFlushTask;
+
+    private final AtomicLong executionIdGen = new AtomicLong();
+
+    private BulkRequest bulkRequest;
+    private final BulkRequestHandler2 bulkRequestHandler;
+    private final Runnable onClose;
+
+    private volatile boolean closed = false;
+    /*
+     * This lock is used to protect two things related to the bulkRequest object: (1) it makes sure that two threads do not add requests
+     * to the BulkRequest at the same time since BulkRequest is not threadsafe and (2) it makes sure that no other thread is writing to
+     * the BulkRequest when we swap the bulkRequest variable over to a new BulkRequest object.
+     */
+    private final ReentrantLock lock = new ReentrantLock();
```

Comment:
Do you need a `ReentrantLock` here? I think it could just be a plain `Object` which you use as a mutex in a `synchronized (mutex) {...}` block.

Reply:
Synchronized would work fine here now that all this is doing is adding to a BulkRequest / adding a BulkRequest to a queue. Is there an advantage to it though? Performance?

Reply:
I just find it less noisy, no need for those `finally` blocks (or for reviewers to check that they're all there). I spent at least a little time trying to understand if we were using anything that makes `ReentrantLock` different from a regular mutex. AFAICT we aren't.

Reply:
Yeah it was a carryover from the old code -- we definitely needed it then (although it was buggy).